File: lib/domain/recovery/save_buffer_to_recovery.dart
import 'dart:io';

import 'package:buffer/domain/recovery/recovery_repository.dart';

/// Use case: save the current buffer text to the recovery store.
///
/// ## Trim-guard (EC-M2-02, §5.1.2)
///
/// When [text] is empty or whitespace-only after trimming, returns `null`
/// immediately — no filesystem I/O is performed. This guard lives HERE at the
/// use-case boundary, not in [RecoveryRepository] and not gated on any
/// `isDirty` flag.
///
/// The RAW (un-trimmed) [text] is passed to [RecoveryRepository.save] — only
/// the empty/non-empty DECISION uses trim.
///
/// ## Return value
///
/// Returns `null` when nothing was saved (empty/whitespace input).
/// Returns the written [File] when the save succeeds.
///
/// ## Error contract
///
/// A [FileSystemException] from the repository propagates to the caller
/// unchanged; the lifecycle host (EC-M2-08) is responsible for catching it.
class SaveBufferToRecovery {
  const SaveBufferToRecovery(this._repository);

  final RecoveryRepository _repository;

  /// Saves [text] to the recovery store, or returns `null` if [text] is
  /// empty/whitespace-only.
  ///
  /// On a successful save, `trim(10)` is called to ensure the recovery
  /// directory never retains more than 10 files (FR-M5-03, §5.1.4).
  /// If [RecoveryRepository.save] throws, the exception propagates before
  /// `trim` runs — no catch, lifecycle host is responsible (EC-M2-08).
  Future<File?> call(String text) async {
    if (text.trim().isEmpty) return null;
    final file = await _repository.save(text);
    await _repository.trim(10);
    return file;
  }

  /// Synchronous entry point for the `paused`/`detached` lifecycle path.
  ///
  /// ## Why synchronous (T-03 fix)
  ///
  /// On Android (and iOS), the OS may freeze the Dart isolate immediately
  /// after the [AppLifecycleState.paused] callback returns. Any pending
  /// microtasks — including the continuation of an `async` save — are never
  /// scheduled, so async I/O is silently dropped. Calling [callSync] from
  /// inside [LifecycleBufferHost.didChangeAppLifecycleState] guarantees the
  /// bytes hit disk before the callback returns.
  ///
  /// ## Trim guard (EC-M2-02)
  ///
  /// Returns `null` immediately when [text] is empty or whitespace-only.
  /// The raw [text] (un-trimmed) is passed to
  /// [RecoveryRepository.saveSync], which also performs the synchronous
  /// trim-to-10 step internally (C-05 / NFR-M5-01).
  ///
  /// ## Error contract
  ///
  /// A [FileSystemException] propagates unchanged to the caller. The
  /// lifecycle host (EC-M2-08) wraps the call in try/catch and logs the
  /// error — this use-case method never swallows it.
  File? callSync(String text) {
    if (text.trim().isEmpty) return null;
    // The repository performs its own synchronous trim-to-10 after writing;
    // we do NOT also call trim() here (no async trim on the sync path).
    return _repository.saveSync(text);
  }
}

File: shared/src/commonMain/kotlin/com/paolosantucci/foglietto/shared/recovery/SaveBufferToRecovery.kt
package com.paolosantucci.foglietto.shared.recovery

/**
 * Use case: save the current buffer text to the recovery store.
 *
 * Port of the Dart `SaveBufferToRecovery` class (main:lib/domain/recovery/save_buffer_to_recovery.dart).
 *
 * ## Trim-guard (FR-07, EC-06)
 *
 * When [text] is empty or whitespace-only after trimming, returns `null`
 * immediately — **zero** repository calls are made (neither [RecoveryRepository.save]
 * nor [RecoveryRepository.trim] is invoked). This guard lives HERE at the
 * use-case boundary, not inside [RecoveryRepository].
 *
 * The RAW (un-trimmed) [text] is passed to [RecoveryRepository.save] — only
 * the empty/non-empty DECISION uses `trim()`.
 *
 * ## Call order (FR-08)
 *
 * When text is non-empty:
 *   1. `repository.save(rawText)` — the literal, un-trimmed text.
 *   2. `repository.trim(10)`       — the keep literal lives HERE, not on the repo.
 *
 * ## Error contract (FR-09, EC-05)
 *
 * If `save` throws, the exception propagates to the caller unchanged.
 * `trim` is **not** called when `save` throws.
 *
 * ## Recovery is always on (FR-10, R-A6)
 *
 * There is no `emergencyRecoveryEnabled` gate, toggle, or setting in this class.
 * Recovery is unconditionally active.
 *
 * ## Return value
 *
 * Returns `null` when nothing was saved (empty/whitespace-only input).
 * Returns the written file path [String] when the save succeeds.
 *
 * Spec refs: §5.1.1, §4.2, FR-07, FR-08, FR-09, FR-10, EC-05, EC-06.
 */
class SaveBufferToRecovery(private val repository: RecoveryRepository) {

    /**
     * Saves [text] to the recovery store, or returns `null` if [text] is
     * empty/whitespace-only.
     *
     * Delegates the raw, un-trimmed [text] to [RecoveryRepository.save], then
     * calls `repository.trim(10)` to cap the store at 10 files. The literal
     * `10` is the use-case boundary — it does not live on [RecoveryRepository].
     *
     * If `save` throws, the exception propagates unchanged and `trim` is never
     * called.
     */
    operator fun invoke(text: String): String? {
        if (text.trim().isEmpty()) return null
        val path = repository.save(text)
        repository.trim(10)
        return path
    }
}
